tools/chpldef/compute-goto-entity.cpp

```diff
@@ -0,0 +1,263 @@
+ * limitations under the License.
+ */
+
+#include "./Message.h"
+#include "./Server.h"
+#include "chpl/framework/query-impl.h"
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/resolution/resolution-types.h"
+
+namespace chpldef {
+
+static const chpl::uast::BuilderResult&
+parseFromPath(chpl::Context* chapel, chpl::UniqueString path) {
+  using namespace chpl::parsing;
+  chpl::UniqueString empty;
+  auto& ret = parseFileToBuilderResultAndCheck(chapel, path, empty);
+  return ret;
+}
+
+namespace {
+  using LineToIdsMap = std::map<int, std::vector<chpl::ID>>;
+}
+
+static bool isSymbolOfInterest(const chpl::uast::AstNode* ast) {
+  if (ast->isIdentifier()) return true;
+  if (ast->isDot()) return true;
+  return false;
+}
+
+// Preorder traversal to collect identifiers in lexical order.
+static void doMapLinesInModule(chpl::Context* chapel,
+                               const chpl::uast::BuilderResult& br,
+                               LineToIdsMap& m,
+                               const chpl::uast::AstNode* node) {
+  for (auto ast : node->children()) {
+    if (ast->isComment()) continue;
+
+    if (isSymbolOfInterest(ast)) {
+      auto loc = br.idToLocation(ast->id(), chpl::UniqueString());
+      CHPL_ASSERT(!loc.isEmpty());
+      auto& v = m[loc.firstLine()];
+      v.push_back(ast->id());
+    }
+
+    bool recurse = !ast->isModule() && !ast->isLeaf();
+    if (recurse) {
+      doMapLinesInModule(chapel, br, m, ast);
+    }
+  }
+}
+
+static LineToIdsMap
+mapLinesToIdsInModule(chpl::Context* chapel, chpl::UniqueString path) {
+  auto& br = parseFromPath(chapel, path);
+  LineToIdsMap ret;
+  auto it = br.topLevelExpressions();
+  for (auto ast : it) doMapLinesInModule(chapel, br, ret, ast);
+  return ret;
+}
+
+static const LineToIdsMap&
+mapLinesToIdsInModuleQuery(chpl::Context* chapel, chpl::UniqueString path) {
+  using namespace chpl;
+  QUERY_BEGIN(mapLinesToIdsInModuleQuery, chapel, path);
+  auto ret = mapLinesToIdsInModule(chapel, path);
+  return QUERY_END(ret);
+}
+
+static chpl::Location
+chplLocFromUriAndPos(chpl::Context* chapel, const std::string& uri,
+                     Position pos) {
+  auto path = chpl::UniqueString::get(chapel, uri);
+  int l1 = pos.line + 1;
+  int c1 = pos.character + 1;
+  int l2 = pos.line + 1;
+  int c2 = pos.character + 1;
+  return chpl::Location(path, l1, c1, l2, c2);
+}
+
+static const chpl::uast::AstNode*
+chooseAstClosestToLocation(chpl::Context* chapel,
+                           const chpl::uast::AstNode* astOld,
+                           const chpl::uast::AstNode* astNew,
+                           chpl::Location loc) {
+  using namespace chpl::parsing;
+  CHPL_ASSERT(astNew);
+  CHPL_ASSERT(astOld != astNew);
+
+  auto locOld = astOld ? locateAst(chapel, astOld) : chpl::Location();
+  auto locNew = locateAst(chapel, astNew);
+  if (!locNew.contains(loc)) return astOld;
+  if (!astOld) return astNew;
```

Comment:
Noting that this will run `locateAst`, possibly on the same AST node, `O(n)` times where `n` is the number of identifiers on a line.